src/System.Sorting/src/TimSort.cs

```diff
@@ -0,0 +1,680 @@
+                    if (order > 0)
+                    {
+                        hi = mid;
+                        //lo = mid + 1;
+                    }
+                    else
+                    {
+                        lo = mid + 1;
+                    }
+                }
+                               
+                int moveBy = value - lo;
+                Array.Copy(array, lo, array, lo+1, moveBy);
+                array[lo] = pivot;
+            }
+        }
+
+        /// <summary>
+		/// Pushes the specified run onto the pending-run stack.
+		/// </summary>
+		/// <param name="runBase">index of the first element in the run.</param>
+		/// <param name="runLength">the number of elements in the run.</param>
+        private void PushRun(int runBase, int runLength)
+        {
+            arrayRunBase[stackSize] = runBase;
+            arrayRunLength[stackSize] = runLength;
+            stackSize++;
+        }
+
+        /// <summary>
+		/// Examines the stack of runs waiting to be merged and merges adjacent runs until the stack invariants are
+		/// reestablished: 
+		/// <c><![CDATA[1. runLen[i - 3] > runLen[i - 2] + runLen[i - 1] ]]></c> and 
+		/// <c><![CDATA[2. runLen[i - 2] > runLen[i - 1] ]]></c>
+		/// This method is called each time a new run is pushed onto the stack		
+		/// </summary>
+        private void MergeCollapse()
+        {
+            while (stackSize > 1)
+            {
+                int n = stackSize - 2;
+
+                //runLen[i-1]<run[i-2]+runLen[i-3] as an outcome of condition mentioned above.
+                if (n > 0 && arrayRunLength[n - 1] <= arrayRunLength[n] + arrayRunLength[n + 1])
+                {          
+                    //Merge runLen[i-2] with runLen[i-3] or runLen[i-1] depending on which is smaller
+                    if (arrayRunLength[n - 1] < arrayRunLength[n + 1]) n--;
+                    MergeAt(n);
+                }
+                else if (arrayRunLength[n] <= arrayRunLength[n + 1])
+                {
+                    MergeAt(n);
+                }
+                else
+                {
+                    break; 
+                }
+            }
+        }
+
+        /// <summary>
+        /// Merges all runs on the stack until only one remains.  This method is called once, to complete the sort.
+        /// </summary>
+        private void MergeForceCollapse()
+        {
+            while (stackSize > 1)
+            {
+                int n = stackSize - 2;
+                if (n > 0 && arrayRunLength[n - 1] < arrayRunLength[n + 1]) n--;
+                MergeAt(n);
+            }
+        }
+
+        /// <summary>
+        /// Merges the two runs at stack indices i and i+1.  Run i must be the penultimate or antepenultimate run on the stack. 
+        /// In other words, i must be equal to stackSize-2 or stackSize-3.
+        /// </summary>
+        /// <param name="runIndex">stack index of the first of the two runs to merge.</param>
```

Comment:
I assume that `runIndex` is the same thing as `i` in the above comment?  It would be helpful to update the comment if so.